File: src/chess_ml_coach/report.py
from __future__ import annotations

import math
from dataclasses import dataclass

import chess
import pandas as pd

MATE_THRESHOLD_CP = 50_000

DISPLAY_LABELS = {
    "samples": "Moves",
    "mistake_rate": "Mistake rate",
    "blunder_rate": "Blunder rate",
    "median_cpl": "Typical eval loss",
    "mean_cpl_non_mate": "Average eval loss",
    "mate_blunders": "Mate mistakes",
    "dimension": "Category",
    "context": "Context",
    "vs_baseline": "Vs baseline",
    "game_id": "Game ID",
    "game": "Game",
    "move": "Move",
    "your_move": "Your move",
    "better_move": "Best move",
    "eval_loss": "Eval loss",
    "quality": "Quality",
    "game_url": "Game link",
    "feature": "Signal",
    "relative_importance": "Relative importance",
}

FEATURE_LABELS = {
    "eval_before_cp": "Position evaluation",
    "engine_eval_before_cp": "Position evaluation",
    "clock_seconds": "Time remaining",
    "legal_move_count": "Legal move choices",
    "fullmove_number": "Move number",
    "rating_difference": "Rating difference",
    "material_balance": "Material advantage",
    "total_non_king_material": "Material remaining",
    "white_rating": "White rating",
    "black_rating": "Black rating",
    "material_balance_white": "White material advantage",
    "white_material": "White material",
    "black_material": "Black material",
    "king_ring_attacks": "King pressure",
    "own_castling_rights": "Castling options",
    "own_doubled_pawns": "Doubled pawns",
    "own_isolated_pawns": "Isolated pawns",
    "own_pawn_islands": "Pawn islands",
}


@dataclass(frozen=True)
class CoachingReport:
    overall: dict[str, float | int]
    by_color: pd.DataFrame
    by_phase: pd.DataFrame
    by_opening: pd.DataFrame
    by_time_control: pd.DataFrame
    recurring_contexts: pd.DataFrame
    candidate_positions: pd.DataFrame
    feature_importance: pd.DataFrame


def _delivered_checkmate(value: object) -> bool:
    if value is None or pd.isna(value):
        return False
    try:
        return chess.Board(str(value)).is_checkmate()
    except ValueError:
        return False


def _sanitize_analysis(frame: pd.DataFrame) -> pd.DataFrame:
    clean = frame.copy()
    same_move = pd.Series(False, index=clean.index)
    if {"uci", "best_move_uci"}.issubset(clean.columns):
        actual = clean["uci"].fillna("").astype(str)
        best = clean["best_move_uci"].fillna("").astype(str)
        same_move = actual.ne("") & best.ne("") & actual.eq(best)

    delivered_mate = pd.Series(False, index=clean.index)
    if "fen_after" in clean.columns:
        delivered_mate = clean["fen_after"].map(_delivered_checkmate)

    definitely_good = same_move | delivered_mate
    if definitely_good.any():
        clean.loc[definitely_good, "cpl"] = 0
        clean.loc[definitely_good, "quality"] = "good"
        clean.loc[definitely_good, "significant_mistake"] = 0
    return clean


def _mate_mask(frame: pd.DataFrame) -> pd.Series:
    mask = pd.Series(False, index=frame.index)
    if "eval_before_cp" in frame.columns:
        mask |= pd.to_numeric(frame["eval_before_cp"], errors="coerce").abs() >= MATE_THRESHOLD_CP
    if "eval_after_cp" in frame.columns:
        mask |= pd.to_numeric(frame["eval_after_cp"], errors="coerce").abs() >= MATE_THRESHOLD_CP
    if "cpl" in frame.columns:
        mask |= pd.to_numeric(frame["cpl"], errors="coerce") >= MATE_THRESHOLD_CP
    return mask.fillna(False)


def _summary_stats(frame: pd.DataFrame) -> dict[str, float | int]:
    mate_mask = _mate_mask(frame)
    normal_cpl = pd.to_numeric(frame.loc[~mate_mask, "cpl"], errors="coerce").dropna()
    return {
        "samples": len(frame),
        "mistake_rate": float(frame["significant_mistake"].mean()),
        "blunder_rate": float((frame["quality"] == "blunder").mean()),
        "median_cpl": float(normal_cpl.median()) if not normal_cpl.empty else 0.0,
        "mean_cpl_non_mate": float(normal_cpl.mean()) if not normal_cpl.empty else 0.0,
        "mate_blunders": int((mate_mask & (frame["quality"] == "blunder")).sum()),
    }


def _aggregate(frame: pd.DataFrame, column: str, min_group_size: int) -> pd.DataFrame:
    if column not in frame.columns:
        return pd.DataFrame(
            columns=[
                "samples",
                "mistake_rate",
                "blunder_rate",
                "median_cpl",
                "mean_cpl_non_mate",
                "mate_blunders",
            ]
        )
    rows: list[dict] = []
    labels: list[object] = []
    for label, group in frame.groupby(column, dropna=False):
        stats = _summary_stats(group)
        if int(stats["samples"]) < min_group_size:
            continue
        labels.append(label)
        rows.append(stats)
    result = pd.DataFrame(rows, index=labels)
    result.index.name = column
    if result.empty:
        return result
    return result.sort_values(["mistake_rate", "samples"], ascending=[False, False])


def _humanize_opening_name(value: object) -> str:
    text = str(value).strip()
    replacements = [
        ("Kings Pawn", "King's Pawn"),
        ("Queens Pawn", "Queen's Pawn"),
        ("Kings Indian", "King's Indian"),
        ("Queens Indian", "Queen's Indian"),
        ("Queens Gambit", "Queen's Gambit"),
        ("Petrovs Defense", "Petrov's Defense"),
    ]
    for old, new in replacements:
        text = text.replace(old, new)
    return text


def _opening_label(row: pd.Series) -> str:
    eco = str(row.get("eco") or "unknown")
    opening = row.get("opening")
    if opening is None or pd.isna(opening) or str(opening).strip().lower() in {"", "unknown", "nan"}:
        return eco
    return f"{_humanize_opening_name(opening)} ({eco})"


def _recurring_contexts(
    overall_mistake_rate: float,
    tables: list[tuple[str, pd.DataFrame]],
) -> pd.DataFrame:
    rows: list[dict] = []
    for dimension, table in tables:
        for context, row in table.iterrows():
            excess = float(row.mistake_rate) - overall_mistake_rate
            score = excess * math.log1p(int(row.samples))
            rows.append(
                {
                    "dimension": dimension,
                    "context": str(context),
                    "samples": int(row.samples),
                    "mistake_rate": float(row.mistake_rate),
                    "blunder_rate": float(row.blunder_rate),
                    "median_cpl": float(row.median_cpl),
                    "vs_baseline": excess,
                    "_score": score,
                }
            )
    if not rows:
        return pd.DataFrame(
            columns=[
                "dimension",
                "context",
                "samples",
                "mistake_rate",
                "blunder_rate",
                "median_cpl",
                "vs_baseline",
            ]
        )
    return (
        pd.DataFrame(rows)
        .sort_values("_score", ascending=False, kind="stable")
        .drop(columns=["_score"])
        .reset_index(drop=True)
    )


def _best_move_san(fen: object, best_move_uci: object) -> str:
    if fen is None or best_move_uci is None or pd.isna(fen) or pd.isna(best_move_uci):
        return "unknown"
    try:
        board = chess.Board(str(fen))
        move = chess.Move.from_uci(str(best_move_uci))
        return board.san(move)
    except (ValueError, AssertionError):
        return str(best_move_uci)


def _move_label(row: pd.Series) -> str:
    fullmove = row.get("fullmove_number")
    san = row.get("san")
    if fullmove is None or pd.isna(fullmove):
        return str(san or row.get("uci") or "unknown")
    separator = "." if row.get("color") == "white" else "..."
    return f"{int(fullmove)}{separator}{san or row.get('uci') or 'unknown'}"


def _game_label(row: pd.Series) -> str:
    white = row.get("white")
    black = row.get("black")
    if white is None or black is None or pd.isna(white) or pd.isna(black):
        return "Game"
    return f"{white} vs {black}"


def _candidate_positions(
    frame: pd.DataFrame,
    limit: int = 20,
    max_per_game: int = 2,
) -> pd.DataFrame:
    mistakes = frame[frame["quality"].isin(["mistake", "blunder"])].copy()
    if {"uci", "best_move_uci"}.issubset(mistakes.columns):
        mistakes = mistakes[
            mistakes["best_move_uci"].isna()
            | mistakes["uci"].isna()
            | (mistakes["uci"].astype(str) != mistakes["best_move_uci"].astype(str))
        ].copy()

    rows: list[dict] = []
    per_game: dict[str, int] = {}
    mate_mask = _mate_mask(mistakes)
    # A stable sort keeps equal losses in chronological order on every platform.
    ranked = mistakes.assign(_mate_related=mate_mask).sort_values(
        "cpl", ascending=False, kind="stable"
    )
    for _, row in ranked.iterrows():
        game_id = str(row["game_id"])
        if per_game.get(game_id, 0) >= max_per_game:
            continue
        cpl = float(row["cpl"])
        rows.append(
            {
                "game_id": game_id,
                "game": _game_label(row),
                "move": _move_label(row),
                "your_move": str(row.get("san") or row.get("uci") or "unknown"),
                "better_move": _best_move_san(row.get("fen_before"), row.get("best_move_uci")),
                "eval_loss": "Mate swing" if bool(row["_mate_related"]) else f"{cpl / 100:.2f} pawns",
                "quality": str(row.get("quality", "unknown")),
                "game_url": str(row.get("source_url") or ""),
            }
        )
        per_game[game_id] = per_game.get(game_id, 0) + 1
        if len(rows) >= limit:
            break
    return pd.DataFrame(
        rows,
        columns=[
            "game_id",
            "game",
            "move",
            "your_move",
            "better_move",
            "eval_loss",
            "quality",
            "game_url",
        ],
    )


def _human_feature_name(raw: object) -> str:
    name = str(raw)
    if "__" in name:
        name = name.split("__", 1)[1]
    if name in FEATURE_LABELS:
        return FEATURE_LABELS[name]

    categorical_prefixes = {
        "game_phase_": "Game phase",
        "time_control_category_": "Time control",
        "color_": "Color",
        "eco_": "Opening code",
        "opening_": "Opening",
    }
    for prefix, label in categorical_prefixes.items():
        if name.startswith(prefix):
            value = name[len(prefix) :].replace("_", " ")
            return f"{label}: {value}"

    return name.replace("_", " ").strip().title()


def _feature_importance_frame(feature_importance: list[dict] | None) -> pd.DataFrame:
    raw = pd.DataFrame(feature_importance or [], columns=["feature", "importance"])
    if raw.empty:
        return pd.DataFrame(columns=["feature", "relative_importance"])
    numeric = pd.to_numeric(raw["importance"], errors="coerce").fillna(0.0)
    total = float(numeric.sum())
    relative = numeric / total if total > 0 else numeric * 0
    return pd.DataFrame(
        {
            "feature": raw["feature"].map(_human_feature_name),
            "relative_importance": relative,
        }
    )


def build_coaching_report(
    frame: pd.DataFrame,
    min_group_size: int = 10,
    *,
    feature_importance: list[dict] | None = None,
) -> CoachingReport:
    if frame.empty:
        raise ValueError("Feature dataset is empty")
    clean = _sanitize_analysis(frame)
    overall = _summary_stats(clean)
    overall_mistake_rate = float(overall["mistake_rate"])
    by_color = _aggregate(clean, "color", min_group_size)
    by_phase = _aggregate(clean, "game_phase", min_group_size)

    opening_frame = clean.copy()
    opening_frame["opening_display"] = opening_frame.apply(_opening_label, axis=1)
    by_opening = _aggregate(opening_frame, "opening_display", min_group_size)
    by_opening.index.name = "opening"

    by_time_control = _aggregate(clean, "time_control_category", min_group_size)
    contexts = _recurring_contexts(
        overall_mistake_rate,
        [
            ("color", by_color),
            ("phase", by_phase),
            ("opening", by_opening),
            ("time_control", by_time_control),
        ],
    )
    return CoachingReport(
        overall=overall,
        by_color=by_color,
        by_phase=by_phase,
        by_opening=by_opening,
        by_time_control=by_time_control,
        recurring_contexts=contexts,
        candidate_positions=_candidate_positions(clean),
        feature_importance=_feature_importance_frame(feature_importance),
    )


def _format_cell(column: str, value: object) -> str:
    if pd.isna(value):
        return "n/a"
    if column in {"mistake_rate", "blunder_rate", "relative_importance"}:
        return f"{float(value):.1%}"
    if column == "vs_baseline":
        return f"{float(value) * 100:+.1f} pp"
    if column in {"median_cpl", "mean_cpl_non_mate"}:
        return f"{float(value) / 100:.2f} pawns"
    if column == "game_url":
        text = str(value)
        return f"[Open game]({text})" if text else "n/a"
    if column in {"quality", "dimension"}:
        return str(value).replace("_", " ").title()
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def _table_markdown(
    frame: pd.DataFrame,
    *,
    include_index: bool = True,
    max_rows: int = 12,
) -> str:
    limited = frame.head(max_rows).copy()
    if limited.empty:
        return "_Insufficient sample size._"
    if include_index:
        index_name = limited.index.name or "group"
        columns = [index_name, *limited.columns.tolist()]
        data_rows = [[idx, *row.tolist()] for idx, row in limited.iterrows()]
    else:
        columns = limited.columns.tolist()
        data_rows = limited.values.tolist()

    display_columns = [DISPLAY_LABELS.get(column, str(column).replace("_", " ").title()) for column in columns]
    header = "| " + " | ".join(display_columns) + " |"
    separator = "| " + " | ".join("---" for _ in columns) + " |"
    body = []
    for row in data_rows:
        body.append(
            "| "
            + " | ".join(_format_cell(column, value) for column, value in zip(columns, row))
            + " |"
        )
    return "\n".join([header, separator, *body])


def _context_title(dimension: str, context: str) -> str:
    if dimension in {"color", "phase", "time_control"}:
        return context.replace("_", " ").title()
    return context


def _priority_lines(report: CoachingReport) -> list[str]:
    positive = report.recurring_contexts[report.recurring_contexts["vs_baseline"] > 0].head(3)
    if positive.empty:
        return ["- No recurring context is clearly worse than your overall baseline yet."]
    lines: list[str] = []
    for _, row in positive.iterrows():
        title = _context_title(str(row["dimension"]), str(row["context"]))
        lines.append(
            f"- **{title}**: {float(row['mistake_rate']):.1%} significant mistakes "
            f"across {int(row['samples'])} moves "
            f"({float(row['vs_baseline']) * 100:+.1f} percentage points vs your baseline)."
        )
    return lines


def render_markdown(report: CoachingReport) -> str:
    candidates = report.candidate_positions.drop(columns=["game_id"], errors="ignore")
    lines = [
        "# Chess ML Coach Report",
        "",
        "## Your priorities",
        *_priority_lines(report),
        "",
        "Use these as training priorities, not absolute judgments; larger samples are more reliable.",
        "",
        "## Overall",
        f"- Analyzed moves: {report.overall['samples']}",
        f"- Mistake rate: {float(report.overall['mistake_rate']):.1%}",
        f"- Blunder rate: {float(report.overall['blunder_rate']):.1%}",
        f"- Typical eval loss: {float(report.overall['median_cpl']) / 100:.2f} pawns",
        f"- Average eval loss: {float(report.overall['mean_cpl_non_mate']) / 100:.2f} pawns",
        f"- Mate mistakes: {report.overall['mate_blunders']}",
        "",
        "## By color",
        _table_markdown(report.by_color),
        "",
        "## By phase",
        _table_markdown(report.by_phase),
        "",
        "## Openings",
        _table_markdown(report.by_opening),
        "",
        "## Time controls",
        _table_markdown(report.by_time_control),
        "",
        "## Recurring weakness contexts",
        _table_markdown(report.recurring_contexts, include_index=False),
        "",
        "## Candidate training positions",
        "These are positions from your own games to review first.",
        "",
        _table_markdown(candidates, include_index=False, max_rows=20),
    ]
    if not report.feature_importance.empty:
        lines.extend(
            [
                "",
                "## Model feature importance",
                "Feature importance is associative, not causal.",
                "",
                _table_markdown(report.feature_importance, include_index=False),
            ]
        )
    return "\n".join(lines).rstrip() + "\n"

File: web/src/analysis/report.ts
// Coaching report ported from report.py (report schema 1). Output markdown must match
// render_markdown(build_coaching_report(...)) byte for byte; see parity.test.ts.
import { Chess } from 'chess.js'

import { parseUci } from './board'
import { MATE_THRESHOLD_CP } from './classify'
import type { FeatureRow } from './features'
import { formatFixed, formatPercent, formatSigned, pyTitle } from './pyformat'

type Cell = string | number | null
type Stats = {
  samples: number
  mistake_rate: number
  blunder_rate: number
  median_cpl: number
  mean_cpl_non_mate: number
  mate_blunders: number
}
type ReportRow = { quality: string; significant_mistake: number; cpl: number; mate: boolean }
export type GroupTable = { indexName: string; rows: { label: string | null; stats: Stats }[] }
export type ContextRow = {
  dimension: string
  context: string
  samples: number
  mistake_rate: number
  blunder_rate: number
  median_cpl: number
  vs_baseline: number
}
export type CandidateRow = {
  game_id: string
  game: string
  move: string
  your_move: string
  better_move: string
  eval_loss: string
  quality: string
  game_url: string
}
export type CoachingReport = {
  overall: Stats
  by_color: GroupTable
  by_phase: GroupTable
  by_opening: GroupTable
  by_time_control: GroupTable
  recurring_contexts: ContextRow[]
  candidate_positions: CandidateRow[]
  feature_importance: { feature: string; relative_importance: number }[]
}

const DISPLAY_LABELS: Record<string, string> = {
  samples: 'Moves', mistake_rate: 'Mistake rate', blunder_rate: 'Blunder rate',
  median_cpl: 'Typical eval loss', mean_cpl_non_mate: 'Average eval loss',
  mate_blunders: 'Mate mistakes', dimension: 'Category', context: 'Context',
  vs_baseline: 'Vs baseline', game_id: 'Game ID', game: 'Game', move: 'Move',
  your_move: 'Your move', better_move: 'Best move', eval_loss: 'Eval loss',
  quality: 'Quality', game_url: 'Game link', feature: 'Signal',
  relative_importance: 'Relative importance',
}
const FEATURE_LABELS: Record<string, string> = {
  eval_before_cp: 'Position evaluation', engine_eval_before_cp: 'Position evaluation',
  clock_seconds: 'Time remaining', legal_move_count: 'Legal move choices',
  fullmove_number: 'Move number', rating_difference: 'Rating difference',
  material_balance: 'Material advantage', total_non_king_material: 'Material remaining',
  white_rating: 'White rating', black_rating: 'Black rating',
  material_balance_white: 'White material advantage', white_material: 'White material',
  black_material: 'Black material', king_ring_attacks: 'King pressure',
  own_castling_rights: 'Castling options', own_doubled_pawns: 'Doubled pawns',
  own_isolated_pawns: 'Isolated pawns', own_pawn_islands: 'Pawn islands',
}
const STAT_COLUMNS = ['samples', 'mistake_rate', 'blunder_rate', 'median_cpl', 'mean_cpl_non_mate', 'mate_blunders'] as const

function deliveredCheckmate(fen: string | null): boolean {
  if (!fen) return false
  try {
    return new Chess(fen).isCheckmate()
  } catch {
    return false
  }
}

type CleanRow = FeatureRow & { mate: boolean }

function sanitize(rows: FeatureRow[]): CleanRow[] {
  return rows.map((row) => {
    const sameMove = !!row.uci && !!row.best_move_uci && row.uci === row.best_move_uci
    const clean = sameMove || deliveredCheckmate(row.fen_after)
      ? { ...row, cpl: 0, quality: 'good', significant_mistake: 0 }
      : { ...row }
    const mate = Math.abs(clean.eval_before_cp) >= MATE_THRESHOLD_CP
      || Math.abs(clean.eval_after_cp) >= MATE_THRESHOLD_CP
      || clean.cpl >= MATE_THRESHOLD_CP
    return { ...clean, mate }
  })
}

function median(values: number[]): number {
  const ordered = [...values].sort((a, b) => a - b)
  const middle = Math.floor(ordered.length / 2)
  return ordered.length % 2 ? ordered[middle] : (ordered[middle - 1] + ordered[middle]) / 2
}

function summaryStats(rows: ReportRow[]): Stats {
  const normal = rows.filter((row) => !row.mate).map((row) => row.cpl)
  const count = rows.length
  return {
    samples: count,
    mistake_rate: rows.reduce((total, row) => total + row.significant_mistake, 0) / count,
    blunder_rate: rows.filter((row) => row.quality === 'blunder').length / count,
    median_cpl: normal.length ? median(normal) : 0,
    mean_cpl_non_mate: normal.length ? normal.reduce((total, value) => total + value, 0) / normal.length : 0,
    mate_blunders: rows.filter((row) => row.mate && row.quality === 'blunder').length,
  }
}

function aggregate(rows: CleanRow[], key: (row: CleanRow) => string | null, indexName: string, minGroupSize: number): GroupTable {
  // pandas groupby(dropna=False) sorts keys with missing values last.
  const groups = new Map<string | null, CleanRow[]>()
  rows.forEach((row) => {
    const label = key(row)
    groups.set(label, [...(groups.get(label) ?? []), row])
  })
  const labels = [...groups.keys()].sort((a, b) => {
    if (a === b) return 0
    if (a === null) return 1
    if (b === null) return -1
    return a < b ? -1 : 1
  })
  const table = labels
    .map((label) => ({ label, stats: summaryStats(groups.get(label)!) }))
    .filter((row) => row.stats.samples >= minGroupSize)
  // sort_values(["mistake_rate", "samples"], ascending=False) is a stable lexsort.
  const ordered = table
    .map((row, index) => ({ row, index }))
    .sort((a, b) => b.row.stats.mistake_rate - a.row.stats.mistake_rate
      || b.row.stats.samples - a.row.stats.samples || a.index - b.index)
    .map(({ row }) => row)
  return { indexName, rows: ordered }
}

function humanizeOpening(value: string): string {
  const replacements: [string, string][] = [
    ['Kings Pawn', "King's Pawn"], ['Queens Pawn', "Queen's Pawn"], ['Kings Indian', "King's Indian"],
    ['Queens Indian', "Queen's Indian"], ['Queens Gambit', "Queen's Gambit"], ['Petrovs Defense', "Petrov's Defense"],
  ]
  let text = value.trim()
  for (const [from, to] of replacements) text = text.split(from).join(to)
  return text
}

function openingLabel(row: CleanRow): string {
  const eco = String(row.eco || 'unknown')
  const opening = row.opening
  if (opening === null || ['', 'unknown', 'nan'].includes(String(opening).trim().toLowerCase())) return eco
  return `${humanizeOpening(opening)} (${eco})`
}

function recurringContexts(overallRate: number, tables: [string, GroupTable][]): ContextRow[] {
  const rows: (ContextRow & { score: number; index: number })[] = []
  for (const [dimension, table] of tables) {
    for (const { label, stats } of table.rows) {
      const excess = stats.mistake_rate - overallRate
      rows.push({
        dimension,
        context: label === null ? 'nan' : label,
        samples: stats.samples,
        mistake_rate: stats.mistake_rate,
        blunder_rate: stats.blunder_rate,
        median_cpl: stats.median_cpl,
        vs_baseline: excess,
        score: excess * Math.log1p(stats.samples),
        index: rows.length,
      })
    }
  }
  return rows
    .sort((a, b) => b.score - a.score || a.index - b.index)
    .map(({ score: _score, index: _index, ...row }) => row)
}

function bestMoveSan(fen: string, bestMoveUci: string | null): string {
  if (!bestMoveUci) return 'unknown'
  try {
    return new Chess(fen).move(parseUci(bestMoveUci)).san
  } catch {
    return bestMoveUci
  }
}

function candidatePositions(rows: CleanRow[], limit = 20, maxPerGame = 2): CandidateRow[] {
  const mistakes = rows
    .filter((row) => row.quality === 'mistake' || row.quality === 'blunder')
    .filter((row) => !row.best_move_uci || !row.uci || row.uci !== row.best_move_uci)
  const ranked = mistakes
    .map((row, index) => ({ row, index }))
    .sort((a, b) => b.row.cpl - a.row.cpl || a.index - b.index)
    .map(({ row }) => row)
  const result: CandidateRow[] = []
  const perGame = new Map<string, number>()
  for (const row of ranked) {
    if ((perGame.get(row.game_id) ?? 0) >= maxPerGame) continue
    const separator = row.color === 'white' ? '.' : '...'
    result.push({
      game_id: row.game_id,
      game: row.white !== null && row.black !== null ? `${row.white} vs ${row.black}` : 'Game',
      move: `${row.fullmove_number}${separator}${row.san || row.uci || 'unknown'}`,
      your_move: String(row.san || row.uci || 'unknown'),
      better_move: bestMoveSan(row.fen_before, row.best_move_uci),
      eval_loss: row.mate ? 'Mate swing' : `${formatFixed(row.cpl / 100, 2)} pawns`,
      quality: String(row.quality ?? 'unknown'),
      game_url: String(row.source_url || ''),
    })
    perGame.set(row.game_id, (perGame.get(row.game_id) ?? 0) + 1)
    if (result.length >= limit) break
  }
  return result
}

function humanFeatureName(raw: string): string {
  let name = raw
  if (name.includes('__')) name = name.split('__').slice(1).join('__')
  if (name in FEATURE_LABELS) return FEATURE_LABELS[name]
  const prefixes: [string, string][] = [
    ['game_phase_', 'Game phase'], ['time_control_category_', 'Time control'], ['color_', 'Color'],
    ['eco_', 'Opening code'], ['opening_', 'Opening'],
  ]
  for (const [prefix, label] of prefixes) {
    if (name.startsWith(prefix)) return `${label}: ${name.slice(prefix.length).split('_').join(' ')}`
  }
  return pyTitle(name.split('_').join(' ').trim())
}

function featureImportance(items: { feature: string; importance: number }[] | null) {
  if (!items || !items.length) return []
  const total = items.reduce((sum, item) => sum + (Number.isFinite(item.importance) ? item.importance : 0), 0)
  return items.map((item) => ({
    feature: humanFeatureName(item.feature),
    relative_importance: total > 0 ? item.importance / total : 0,
  }))
}

export function buildCoachingReport(
  rows: FeatureRow[],
  minGroupSize: number,
  importance: { feature: string; importance: number }[] | null,
): CoachingReport {
  if (!rows.length) throw new Error('Feature dataset is empty')
  const clean = sanitize(rows)
  const overall = summaryStats(clean)
  const byColor = aggregate(clean, (row) => row.color, 'color', minGroupSize)
  const byPhase = aggregate(clean, (row) => row.game_phase, 'game_phase', minGroupSize)
  const byOpening = aggregate(clean, openingLabel, 'opening', minGroupSize)
  const byTimeControl = aggregate(clean, (row) => row.time_control_category, 'time_control_category', minGroupSize)
  return {
    overall,
    by_color: byColor,
    by_phase: byPhase,
    by_opening: byOpening,
    by_time_control: byTimeControl,
    recurring_contexts: recurringContexts(overall.mistake_rate, [
      ['color', byColor], ['phase', byPhase], ['opening', byOpening], ['time_control', byTimeControl],
    ]),
    candidate_positions: candidatePositions(clean),
    feature_importance: featureImportance(importance),
  }
}

function formatCell(column: string, value: Cell): string {
  if (value === null) return 'n/a'
  if (['mistake_rate', 'blunder_rate', 'relative_importance'].includes(column)) return formatPercent(Number(value), 1)
  if (column === 'vs_baseline') return `${formatSigned(Number(value) * 100, 1)} pp`
  if (column === 'median_cpl' || column === 'mean_cpl_non_mate') return `${formatFixed(Number(value) / 100, 2)} pawns`
  if (column === 'game_url') return value ? `[Open game](${value})` : 'n/a'
  if (column === 'quality' || column === 'dimension') return pyTitle(String(value).split('_').join(' '))
  return String(value)
}

function markdownTable(columns: string[], data: Cell[][], maxRows = 12): string {
  const limited = data.slice(0, maxRows)
  if (!limited.length) return '_Insufficient sample size._'
  const header = `| ${columns.map((column) => DISPLAY_LABELS[column] ?? pyTitle(column.split('_').join(' '))).join(' | ')} |`
  const separator = `| ${columns.map(() => '---').join(' | ')} |`
  const body = limited.map((row) => `| ${row.map((value, index) => formatCell(columns[index], value)).join(' | ')} |`)
  return [header, separator, ...body].join('\n')
}

function groupMarkdown(table: GroupTable): string {
  return markdownTable(
    [table.indexName, ...STAT_COLUMNS],
    table.rows.map(({ label, stats }) => [label, ...STAT_COLUMNS.map((column) => stats[column])]),
  )
}

function contextTitle(dimension: string, context: string): string {
  return ['color', 'phase', 'time_control'].includes(dimension) ? pyTitle(context.split('_').join(' ')) : context
}

function priorityLines(report: CoachingReport): string[] {
  const positive = report.recurring_contexts.filter((row) => row.vs_baseline > 0).slice(0, 3)
  if (!positive.length) return ['- No recurring context is clearly worse than your overall baseline yet.']
  return positive.map((row) =>
    `- **${contextTitle(row.dimension, row.context)}**: ${formatPercent(row.mistake_rate, 1)} significant mistakes `
    + `across ${row.samples} moves `
    + `(${formatSigned(row.vs_baseline * 100, 1)} percentage points vs your baseline).`)
}

export function renderMarkdown(report: CoachingReport): string {
  const contextColumns = ['dimension', 'context', 'samples', 'mistake_rate', 'blunder_rate', 'median_cpl', 'vs_baseline'] as const
  const candidateColumns = ['game', 'move', 'your_move', 'better_move', 'eval_loss', 'quality', 'game_url'] as const
  const overall = report.overall
  const lines = [
    '# Chess ML Coach Report',
    '',
    '## Your priorities',
    ...priorityLines(report),
    '',
    'Use these as training priorities, not absolute judgments; larger samples are more reliable.',
    '',
    '## Overall',
    `- Analyzed moves: ${overall.samples}`,
    `- Mistake rate: ${formatPercent(overall.mistake_rate, 1)}`,
    `- Blunder rate: ${formatPercent(overall.blunder_rate, 1)}`,
    `- Typical eval loss: ${formatFixed(overall.median_cpl / 100, 2)} pawns`,
    `- Average eval loss: ${formatFixed(overall.mean_cpl_non_mate / 100, 2)} pawns`,
    `- Mate mistakes: ${overall.mate_blunders}`,
    '',
    '## By color',
    groupMarkdown(report.by_color),
    '',
    '## By phase',
    groupMarkdown(report.by_phase),
    '',
    '## Openings',
    groupMarkdown(report.by_opening),
    '',
    '## Time controls',
    groupMarkdown(report.by_time_control),
    '',
    '## Recurring weakness contexts',
    markdownTable([...contextColumns], report.recurring_contexts.map((row) => contextColumns.map((column) => row[column]))),
    '',
    '## Candidate training positions',
    'These are positions from your own games to review first.',
    '',
    markdownTable([...candidateColumns], report.candidate_positions.map((row) => candidateColumns.map((column) => row[column])), 20),
  ]
  if (report.feature_importance.length) {
    lines.push(
      '',
      '## Model feature importance',
      'Feature importance is associative, not causal.',
      '',
      markdownTable(['feature', 'relative_importance'], report.feature_importance.map((row) => [row.feature, row.relative_importance])),
    )
  }
  return `${lines.join('\n').trimEnd()}\n`
}
